File: src/main/java/net/tassia/parser/ParseException.java
package net.tassia.parser;

/**
 * Thrown when something goes wrong when parsing.
 * If possible, the source string, the position and length of the error are included in this exception.
 * @since Parser 1.0
 * @author Tassilo
 */
public class ParseException extends Exception {

	private final String source;
	private final int position, length;

	/**
	 * Creates a new generic {@link ParseException} that doesn't have any clear source.
	 * @param message the message
	 */
	public ParseException(String message) {
		this(message, null, -1, 0);
	}

	/**
	 * Creates a new {@link ParseException} with the given source and position.
	 * @param message the message
	 * @param source the source string
	 * @param position the position of the error
	 * @param length the length of the error
	 */
	public ParseException(String message, String source, int position, int length) {
		super(message);
		this.source = source;
		this.position = position;
		this.length = length;
	}

	/**
	 * Returns the source string of this error. Can be <code>null</code>.
	 * @return the source string
	 */
	public String getSource() {
		return source;
	}

	/**
	 * Returns the position of the error in the source string. Will be
	 * <code>-1</code> if {@link #getSource()} is <code>null</code>.
	 * @return the position
	 */
	public int getPosition() {
		return position;
	}

	/**
	 * Returns the length of the error in the source string. Will be
	 * <code>0</code> if {@link #getSource()} is <code>null</code>.
	 * @return the length
	 */
	public int getLength() {
		return length;
	}

	/**
	 * Generates an informative display of the error by checking the source, error position and length.
	 * For example:<br/>
	 * <pre>{@code
	 * var ex = new ParseException("Unexpected '@'", "Hell@o World!", 4, 1);
	 * System.out.println(ParseException.getDisplayed(ex));
	 * }</pre>
	 * Would render to:
	 * <pre>
	 * Hell@o World!
	 *     ^
	 * </pre>
	 * @param ex the message to format
	 * @return a displayable string
	 */
	public static String getDisplayed(ParseException ex) {
		var str = ex.source;
		str += "\n";
		str += " ".repeat(ex.position) + "^".repeat(ex.length);
		return str;
	}

}

File: src/main/java/net/tassia/parser/ParseException.kt
package net.tassia.parser

import java.io.PrintWriter
import java.io.StringWriter

/**
 * Thrown when something goes wrong when parsing.
 * If possible, the source string, the position and length
 * of the error will be included in this exception.
 *
 * @param source the source string
 * @param position the error position
 * @param length the error length
 * @param message the message
 * @param cause the cause
 *
 * @since Parser 1.0
 * @author Tassilo
 */
class ParseException(

	/**
	 * The source string of this error.
	 */
	val source: String?,

	/**
	 * The position of the error in the source string.
	 */
	val position: Int,

	/**
	 * The length of the error in the source string.
	 */
	val length: Int,

	/**
	 * The message of the error.
	 */
	message: String? = null,

	/**
	 * The cause of the error.
	 */
	cause: Throwable? = null,

) : Exception(message, cause) {

	/**
	 * Generates an informative display of the error by checking the source, error position and length.
	 *
	 * @return a displayable string
	 */
	fun display(): String {
		var str = source ?: message ?: StringWriter().let {
			printStackTrace(PrintWriter(it))
			return it.toString()
		}

		str += "\n"
		str += " ".repeat(position) + "^".repeat(length)
		return str
	}

}
